File: 30-Day LeetCoding Challenge/Python/productExceptSelf.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        prefix = [1]
        suffix = [1]*(len(nums)+1)
        for i in range(len(nums)):
            prefix.append(prefix[i]*nums[i])
        for i in range(len(nums)-1,-1,-1):
            suffix[i] = suffix[i+1]*nums[i]
        for i in range(len(nums)):
            nums[i] = prefix[i]*suffix[i+1]
        return nums
input = [1,2,3,4]
s = Solution()
print(s.productExceptSelf(input))
                

File: 30-Day LeetCoding Challenge/C++/productExceptSelf.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int total = 1, count_zeros = 0;
        for(int num : nums) {
            if(num != 0)
                total = total*num;
            else count_zeros++;
        }
        vector<int> ans;
        for(int i = 0; i < (int) nums.size(); i++) {
            if(count_zeros == 1 && nums[i] == 0) ans.push_back(total);
            else if(count_zeros > 0) ans.push_back(0);
            else ans.push_back(total/nums[i]);
        }
        return ans;
    }
};

int main() {
    vector<int> input = {1,2,3,4};
    Solution s;
    vector<int> output = s.productExceptSelf(input);
    for(int num : output) {
        cout << num << ", ";
    }
    return 0;
}